### PR TITLE
Security management enhancements

### PR DESCRIPTION
**Pull request by dgutierr**
UF-325: Add help icons with tips to workbench permissions
UF-329: Add automatic disabling of illogical combinations in Security perspective

@ederign Can you please verify?

### CODE REVIEW

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/ACLSettingsView.html [hunk 1]
```diff
@@ -1,18 +1,22 @@
 <div class="pull-left" style="width: 100%;">
-    <div class="row" style="margin-bottom: 10px;">
+    <div class="row" style="margin-bottom: 15px;">
```
Comment (ederign) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/ACLSettingsView.html]: could you please move this for ACLSettingsView.css?

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/ACLSettingsView.html [hunk 2]
```diff
@@ -1,18 +1,22 @@
...
         <div class="col-md-2">
             <label id="homePerspectiveLabel" class="panel-title">Home Perspective</label>
-            <span id="homePerspectiveHelp"><i class="pf pficon-help" data-toggle="tooltip" data-placement="left" style="margin-left: 5px;"></i></span>
+            <span id="homePerspectiveHelp" style="font-size: small;">
```
Comment (ederign) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/ACLSettingsView.html]: could you please move this for ACLSettingsView.css?

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/ACLSettingsView.html [hunk 3]
```diff
@@ -1,18 +1,22 @@
...
             <div id="homePerspectiveSelectorPanel"></div>
         </div>
     </div>
-    <div class="row" style="margin-bottom: 10px;">
+    <div class="row" style="margin-bottom: 15px;">
         <div class="col-md-2">
             <label id="priorityLabel" class="panel-title" style="float: left; margin-right: 5px;">Priority</label>
-            <span id="priorityHelp"><i class="pf pficon-help" data-toggle="tooltip" data-placement="left" style="margin-left: 5px;;"></i></span>
+            <span id="priorityHelp" style="font-size: small;">
+                <i class="pf pficon-help" data-toggle="tooltip" data-placement="left" style="margin-left: 5px;;"></i>
```
Comment (ederign) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/ACLSettingsView.html]: same for the other inline styles.

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/BasePermissionNodeEditor.java [hunk 4]
```diff
@@ -143,16 +147,49 @@ public void onChildPermissionChanged(PermissionNodeEditor childEditor, Permissio
 
     }
 
+    protected void onNodePanelWidthChanged() {
+
```
Comment (ederign) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/BasePermissionNodeEditor.java]: This is a empty default implementation? Or this method could be abstract?

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/BasePermissionNodeEditor.java [hunk 5]
```diff
@@ -143,16 +147,49 @@ public void onChildPermissionChanged(PermissionNodeEditor childEditor, Permissio
...
+    }
+
     protected void onPermissionChanged(Permission permission, boolean on) {
+        // Notify the parent editor
```
Comment (ederign) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/BasePermissionNodeEditor.java]: please, extract this comments to methods

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/LeafPermissionNodeEditor.java [hunk 6]
```diff
@@ -101,7 +97,14 @@ public void edit(PermissionNode node) {
                 switchToogle = permissionSwitch;
             }
             initPermissionSwitchToogle(switchToogle, permission);
-            permissionSwitchMap.put(permission, switchToogle);
+            super.registerPermissionSwitch(permission, switchToogle);
+        }
+
+        // Update the switches status according to the inter-dependencies between the permissions
+        super.processAllPermissionDependencies();
```
Comment (ederign) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/LeafPermissionNodeEditor.java]: please move comments to method calls.

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/LeafPermissionNodeEditorView.html [hunk 7]
```diff
@@ -1,6 +1,10 @@
 <div class="leaf-panel">
     <div data-field="nodeNamePanel" style="display: table-cell; vertical-align: middle; height: 30px;">
-        <label data-field="nodeName"></label>
+        <label data-field="nodeName" style="float: left;"></label>
+        <span data-field="nodeNameHelp" style="display: none; vertical-align: middle; font-size: small;">
```
Comment (ederign) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/LeafPermissionNodeEditorView.html]: Please move the styles for css file.

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/LeafPermissionNodeEditorView.java [hunk 8]
```diff
@@ -63,6 +68,8 @@ public void setNodePanelWidth(int width) {
     @Override
     public void setNodeFullName(String name) {
         nodeName.setTitle(name);
+        nodeNameHelp.setTitle(name);
+        nodeNameHelp.getStyle().removeProperty("display");
```
Comment (ederign) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/LeafPermissionNodeEditorView.java]: maybe it's better to add/remove a css class here.

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/MultiplePermissionNodeEditor.java [hunk 9]
```diff
@@ -174,9 +172,17 @@ public void edit(PermissionNode node) {
                 int n = getExceptionNumber(permission);
                 permissionSwitch.setNumberOfExceptions(n);
             });
-            permissionSwitchMap.put(permission, permissionSwitch);
-            view.addPermission(permissionSwitch);
+            super.registerPermissionSwitch(permission, permissionSwitch);
         }
+        // Update the switches status according to the inter-dependencies between the permissions
+        super.processAllPermissionDependencies();
```
Comment (ederign) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/MultiplePermissionNodeEditor.java]: same comment about comments. :)

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/MultiplePermissionNodeEditorView.html [hunk 10]
```diff
@@ -4,6 +4,9 @@
             <h4 class="panel-title">
                 <a data-field="nodeAnchor" data-toggle="collapse" class="collapsed" href="#" aria-expanded="false"
                    style="font-weight: inherit; color: inherit;"></a>
+                <span data-field="nodeAnchorHelp" style="display: none; vertical-align: middle; font-size: small;">
```
Comment (ederign) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/MultiplePermissionNodeEditorView.html]: Please place that at css file.

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/MultiplePermissionNodeEditorView.java [hunk 11]
```diff
@@ -121,6 +126,8 @@ public void setNodePanelWidth(int width) {
     @Override
     public void setNodeFullName(String name) {
         nodeAnchor.setTitle(name);
+        nodeAnchorHelp.setTitle(name);
+        nodeAnchorHelp.getStyle().removeProperty("display");
```
Comment (ederign) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/MultiplePermissionNodeEditorView.java]: Same comment as above

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/PermissionExceptionSwitch.java [hunk 12]
```diff
@@ -70,11 +72,27 @@ public void setOnChange(Command onChange) {
...
+    }
+
+    @Override
+    public void setEnabled(boolean enabled) {
+        view.setEnabled(enabled);
+    }
+
+    @Override
+    public void setNumberOfExceptions(int n) {
+
```
Comment (ederign) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/PermissionExceptionSwitch.java]: why this empty method?

**Comment by ederign** [comment]
@dgutierr could you please take  a look on my comments? Thanks

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/BasePermissionNodeEditor.java (hunk 4 above)
Comment (dgutierr) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/BasePermissionNodeEditor.java]: It is a default impl as we do not force subclasses to implement it.

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/PermissionExceptionSwitch.java (hunk 12 above)
Comment (dgutierr) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/PermissionExceptionSwitch.java]: Any PermissionSwitch widget may have an exception counter.The PermissionExceptionSwitch is an implementation that does not care about the exception counter as it is terminal.

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/ACLSettingsView.html (hunk 1 above)
Comment (dgutierr) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/ACLSettingsView.html]: Done

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/ACLSettingsView.html (hunk 2 above)
Comment (dgutierr) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/ACLSettingsView.html]: Done

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/ACLSettingsView.html (hunk 3 above)
Comment (dgutierr) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/ACLSettingsView.html]: Done

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/BasePermissionNodeEditor.java (hunk 5 above)
Comment (dgutierr) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/BasePermissionNodeEditor.java]: Done

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/LeafPermissionNodeEditor.java (hunk 6 above)
Comment (dgutierr) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/LeafPermissionNodeEditor.java]: Done

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/LeafPermissionNodeEditorView.html (hunk 7 above)
Comment (dgutierr) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/LeafPermissionNodeEditorView.html]: Done

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/LeafPermissionNodeEditorView.java (hunk 8 above)
Comment (dgutierr) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/LeafPermissionNodeEditorView.java]: Done

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/MultiplePermissionNodeEditor.java (hunk 9 above)
Comment (dgutierr) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/MultiplePermissionNodeEditor.java]: Done

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/MultiplePermissionNodeEditorView.html (hunk 10 above)
Comment (dgutierr) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/MultiplePermissionNodeEditorView.html]: Done

File: uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/MultiplePermissionNodeEditorView.java (hunk 11 above)
Comment (dgutierr) [inline uberfire-extensions/uberfire-security/uberfire-security-management/uberfire-widgets-security-management/src/main/java/org/uberfire/ext/security/management/client/widgets/management/editor/acl/node/MultiplePermissionNodeEditorView.java]: Done

**Comment by ederign** [comment]
+1 thanks @dgutierr

**Pull request merged**